econml/dml.py

```diff
@@ -194,85 +269,108 @@ class _DMLCateEstimatorBase(_RLearner):
+                    F = self._featurizer.fit_transform(X) if fitting else self._featurizer.transform(X)
                     XW = hstack([X, W])
                     return cross_product(XW, hstack([np.ones((shape(XW)[0], 1)), F, W]))
                 else:
-                    return hstack([F, W])
+                    return hstack([X, W])
 
-            def fit(self, X, W, Target):
-                self._model.fit(self._combine(X, W), Target)
+            def fit(self, X, W, Target, sample_weight=None):
+                if sample_weight is not None:
+                    self._model.fit(self._combine(X, W), Target, sample_weight=sample_weight)
+                else:
+                    self._model.fit(self._combine(X, W), Target)
 
             def predict(self, X, W):
                 if (not self._is_Y) and discrete_treatment:
-                    return self._model.predict_proba(self._combine(X, W))
+                    return self._model.predict_proba(self._combine(X, W, fitting=False))
                 else:
-                    return self._model.predict(self._combine(X, W))
+                    return self._model.predict(self._combine(X, W, fitting=False))
 
         class FinalWrapper:
             def __init__(self):
                 self._model = clone(model_final, safe=False)
                 self._featurizer = clone(featurizer, safe=False)
 
-            def fit(self, X, T_res, Y_res):
+            def fit(self, X, T_res, Y_res, sample_weight=None, sample_var=None):
                 # Track training dimensions to see if Y or T is a vector instead of a 2-dimensional array
                 self._d_t = shape(T_res)[1:]
                 self._d_y = shape(Y_res)[1:]
-
-                self._model.fit(cross_product(self._featurizer.fit_transform(X), T_res), Y_res)
-
-            def predict(self, X, T_res=None):
-                # create an identity matrix of size d_t (or just a 1-element array if T was a vector)
-                # the nth row will allow us to compute the marginal effect of the nth component of treatment
-                eye = np.eye(self._d_t[0]) if self._d_t else np.array([1])
-                # TODO: Doing this kronecker/reshaping/transposing stuff so that predict can be called
-                #       rather than just using coef_ seems silly, but one benefit is that we can use linear models
-                #       that don't expose a coef_ (e.g. a GridSearchCV over underlying linear models)
-                flat_eye = reshape(eye, (1, -1))
-                XT = reshape(np.kron(flat_eye, self._featurizer.fit_transform(X)),
-                             ((self._d_t[0] if self._d_t else 1) * shape(X)[0], -1))
-                effects = reshape(self._model.predict(XT), (-1,) + self._d_t + self._d_y)
-                if self._d_t and self._d_y:
-                    return transpose(effects, (0, 2, 1))  # need to return as m by d_y by d_t matrix
+                fts = self._combine(X, T_res)
```

Comment:
I got a warning from StatsModelsLinearRegression in utilities saying "Co-variance matrix is undertermined. Inference will be invalid!". It looks like there is a bug for the _combine function if X is None, which will first replace X as np.ones((n,1)) and self._featurizer.fit_transform(X) will change X as shape of n*2 since include_bias=True by default, and finally the cross_product will double the dimension of T_res (when X is actually None). it will cause warning for the inference.

Reply:
Indeed! This is also related to the current interface which blends together the intercept with the features. An alternative interface would be were we have something like: `fit_intercept=True` flag parameter, which adds an intercept to the featurized vector, as opposed to leaving it to the featurizer to handle the addition of the intercept. A user might forget about that. But for now an easier fix to this that simply doesn't add the extra constant would be better.

Reply:
Filed #119 to address this later.